source/agora/node/Registry.d

```diff
@@ -680,6 +687,171 @@ private struct ZoneData
+
+        return 0;
+    }
+
+    /***************************************************************************
+
+         Returns:
+           All registries in persistent storage
+
+    ***************************************************************************/
+
+    public ResultRange getRegistries () @trusted
+    {            
+        return this.db.execute(this.query_registry_get);
+    }
+
+    /***************************************************************************
+
+         Get payload data from persistent storage
+
+         Params:
+           public_key = the public key that was used to register
+                         the network addresses
+
+        Returns:
+            TypedPayload of name registry associated with `public_key`
+
+    ***************************************************************************/
+
+    public TypedPayload getRegistryTypedPayload (PublicKey public_key) @trusted
+    {
+        auto results = this.db.execute(this.query_payload, public_key.toString());
+
+        if (results.empty)
+            return TypedPayload.init;
+
+        // Address loop consumes data, gather following first
+        const TYPE node_type = to!TYPE(results.front["type"].as!(ushort));
+        const ulong sequence = results.front["sequence"].as!(ulong);
+        const Signature signature = Signature.fromString(results.front["signature"].as!(string));
+
+        Address[] addresses;
+        foreach (Row row; results)
+        {
+            addresses ~= row["address"].as!(Address);
+        }
+
+        const RegistryPayload payload =
+        {
+            data:
+            {
+                public_key : public_key,
+                addresses : addresses,
+                seq : sequence
+            },
+            signature: signature
+        };
+
+        // CNAME and A cannot exist at the same time for a node
+        TypedPayload typed_payload =
+        {
+            type: node_type,
+            payload: payload
+        };
+
+        return typed_payload;
+    }
+
+    /***************************************************************************
+
+         Save name registry payload to the persistent storage
+
+         Params:
+           payload = Payload to save
+
+    ***************************************************************************/
+
+    public void addRegistry (TypedPayload payload) @trusted
+    {
+        if (payload.payload.data.addresses.length <= 0)
+            return;
+
+        db.execute(this.query_signature_add, 
+            payload.payload.data.public_key.toString(),
+            payload.payload.signature.toString(),
+            payload.payload.data.seq);
+
+        // There is no need to check for stale addresses 
+        // since `REPLACE INTO` is used and `DELETE` is cascaded.
+        foreach (Address address; payload.payload.data.addresses)
+        {
+            db.execute(this.query_addresses_add,
+                payload.payload.data.public_key.toString(),
```

Comment:
You don't have to use `toString` here. `d2sqlite` uses the `toString(sink)` approach.

## How does it work ?

The common signature for `toString()` is `string toString() const;`, however this has a major downside: It forces immutability of the return, and hence usually means allocation. That's pretty bad, especially if the only intent is to print it to stdout after.

So we often use composition via delegate. We define `toString` as:
```D
void toString (scope void delegate (scope const char[]) @safe sink) const
```

That's a bit more complicated, but it allows us to avoid allocation. The `toString` method can pass "pieces" of the string to `sink` to build the string without needing to allocate. For example:
```D
public struct Person
{
    ///
    public string name;
    
    ///
    public string city;
    
    ///
    void toString (scope void delegate (scope const char[]) @safe sink) const
    {
        sink("Name: ");
        sink(this.name.length ? this.name : "(null)");
        sink("City: ");
        sink(this.city.length ? this.city : "(null)");
    }
}
````

Now, if you call this like this:
```D
import core.stdc.stdio;

void main ()
{
    scope print = (scope const char[] data) {
        printf("%.*s", cast(int) data.length, data.ptr);
    }
    Person inst;
    inst.toString(print);
    printf("\n");
}
```

This will not allocate. Even better, you can easily compose with this approach:
```D
public struct Group
{
    ///
    public Person[] members;
    
    ///
    void toString (scope void delegate (scope const char[]) @safe sink) const
    {
        foreach (const ref p; this.members)
        {
            sink("New person: ");
            p.toString(s); // No need to get a `string` allocated in `Person.toString` just to append it somewhere else!
            sink("\n");
        }
    }
}
```

Now that's a lot of manual work and doesn't work with `int` and co... But luckily, some library functions accept a `sink`, such as `formattedWrite`:
```D
import std.format : formattedWrite;

public struct Group
{
    ///
    public Person[] members;
    
    ///
    void toString (scope void delegate (scope const char[]) @safe sink) const
    {
        // formattedWrite will just pass the data piecemeal to the `sink`, avoiding intermediate allocations
        foreach (const ref p; this.members)
            formattedWrite(sink, "New person: %s\n", p);
    }
}
```

The great thing about this technique is that it works for pretty much everything where composition is needed. We use it for `toString` a lot, but also for hashing (search for `computeHash`, called from [here](https://github.com/bosagora/crypto/blob/b05147f8844ee8a4e4fd6751df69f66ba30001bd/source/agora/crypto/Hash.d#L139-L140) or binary serialization (look up `serialize`, called from [here](https://github.com/bosagora/serialization/blob/1cb6c19d4b51b82e40438c08d27cdcfc9e487ca7/source/agora/serialization/Serializer.d#L408-L410)).

The only downside with this approach is that you have to stick to the delegate's attributes, unless you template the delegate (but this has other issues - story for another time). So in our case, the delegates have a fixed set of attributes which we never diverge from (and from `toString`, that means only `@safe`).